File: packages/go-agentenkit/core/stream.go
package core

import (
	"encoding/json"
	"sync"
	"time"

	"github.com/zendev-sh/goai"
	"github.com/zendev-sh/goai/provider"
)

// usageJSON is the usage shape a CHUNK payload carries. Same field names the
// TypeScript package (AI SDK v4) publishes, so the React client reads both.
func usageJSON(u provider.Usage) map[string]any {
	return map[string]any{
		"promptTokens":      u.InputTokens,
		"completionTokens":  u.OutputTokens,
		"totalTokens":       u.TotalTokens,
		"cachedInputTokens": u.CacheReadTokens,
	}
}

// ChunkPayload renders a goai stream chunk in the AI SDK part shape the
// TypeScript package publishes as CHUNK. A client written against one
// runtime keeps working against the other.
func ChunkPayload(c provider.StreamChunk) map[string]any {
	switch c.Type {
	case provider.ChunkText:
		return map[string]any{"type": "text-delta", "textDelta": c.Text}
	case provider.ChunkReasoning:
		return map[string]any{"type": "reasoning", "textDelta": c.Text}
	case provider.ChunkToolCallStreamStart:
		return map[string]any{"type": "tool-call-streaming-start", "toolCallId": c.ToolCallID, "toolName": c.ToolName}
	case provider.ChunkToolCallDelta:
		return map[string]any{"type": "tool-call-delta", "toolCallId": c.ToolCallID, "toolName": c.ToolName, "argsTextDelta": c.ToolInput}
	case provider.ChunkToolCall:
		var args any
		if json.Valid([]byte(c.ToolInput)) {
			args = json.RawMessage(c.ToolInput)
		} else {
			args = c.ToolInput
		}
		return map[string]any{"type": "tool-call", "toolCallId": c.ToolCallID, "toolName": c.ToolName, "args": args}
	case provider.ChunkToolResult:
		result := jsonOrString(c.Text)
		if c.Error != nil {
			// A tool that failed: the result names the error, the way a
			// denied or unknown tool's result already does, so a client can
			// mark the call failed from the result alone.
			result = MarshalPayload(map[string]any{"error": c.Error.Error()})
		}
		return map[string]any{"type": "tool-result", "toolCallId": c.ToolCallID, "toolName": c.ToolName, "result": result}
	case provider.ChunkStepFinish:
		return map[string]any{"type": "step-finish", "finishReason": string(c.FinishReason), "usage": usageJSON(c.Usage)}
	case provider.ChunkFinish:
		return map[string]any{"type": "finish", "finishReason": string(c.FinishReason), "usage": usageJSON(c.Usage)}
	case provider.ChunkError:
		msg := ""
		if c.Error != nil {
			msg = c.Error.Error()
		}
		return map[string]any{"type": "error", "error": msg}
	}
	return map[string]any{"type": string(c.Type), "text": c.Text}
}

// drainStream reads a goai stream to the end, firing onChunk per part, and
// turns a provider failure into an error.
//
// goai reports a failure (an aborted call included) as an error chunk and
// then ends the stream; Err() is only meaningful once the stream is drained.
// Draining fully first also means onChunk fires for every part that did
// arrive before the failure.
func drainStream(stream *goai.TextStream, onChunk func(provider.StreamChunk)) error {
	for chunk := range stream.Stream() {
		if onChunk != nil {
			onChunk(chunk)
		}
	}
	return stream.Err()
}

// chunkWindow is how long a merged delta is held before it goes out.
const chunkWindow = 50 * time.Millisecond

// chunkBatcher merges token deltas before they go out as events (§2.2).
// Every published event costs a seq, a stored row and a bus send; one per
// token is thousands a reply. Consecutive text (or reasoning) deltas are
// joined into one delta of the same shape, so a client reads them exactly as
// before. A held delta goes out after chunkWindow, or at once when any other
// kind of chunk comes, or on flush: the loop flushes before a step is
// committed, so its text is never published after the step that produced
// it. Sends are made one at a time, in order.
type chunkBatcher struct {
	send  func(map[string]any)
	mu    sync.Mutex
	held  map[string]any
	timer *time.Timer
}

func newChunkBatcher(send func(map[string]any)) *chunkBatcher {
	return &chunkBatcher{send: send}
}

func isDelta(p map[string]any) bool {
	t, _ := p["type"].(string)
	_, text := p["textDelta"].(string)
	return text && (t == "text-delta" || t == "reasoning")
}

// push takes one chunk payload. A delta is held; anything else is sent,
// after what is held, before push returns.
func (b *chunkBatcher) push(p map[string]any) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if isDelta(p) && b.held != nil && b.held["type"] == p["type"] {
		b.held["textDelta"] = b.held["textDelta"].(string) + p["textDelta"].(string)
		return
	}
	b.flushLocked()
	if isDelta(p) {
		held := make(map[string]any, len(p))
		for k, v := range p {
			held[k] = v
		}
		b.held = held
		b.timer = time.AfterFunc(chunkWindow, b.flush)
		return
	}
	b.send(p)
}

// flush sends what is held.
func (b *chunkBatcher) flush() {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.flushLocked()
}

func (b *chunkBatcher) flushLocked() {
	if b.timer != nil {
		b.timer.Stop()
		b.timer = nil
	}
	if b.held != nil {
		held := b.held
		b.held = nil
		b.send(held)
	}
}

File: packages/agentenkit/src/core/stream.ts
/** Drain a `streamText` result, turning a provider failure into a throw.
 *
 *  streamText reports a failure — an aborted call included — as an `error`
 *  part and then ends the stream NORMALLY, while its `text`/`usage`/`response`
 *  promises never settle. Awaiting those without rethrowing hangs the caller
 *  forever, still holding whatever it owns: for a run segment that is the
 *  thread's run lock, which wedges every later message on the thread.
 *
 *  Draining fully first also means `onChunk` fires for every part that did
 *  arrive before the failure. */
export async function drainOrThrow(fullStream: AsyncIterable<unknown>): Promise<void> {
  let streamError: unknown;
  for await (const part of fullStream) {
    if ((part as { type?: string } | null)?.type === 'error' && streamError === undefined) {
      streamError = (part as { error?: unknown }).error;
    }
  }
  if (streamError !== undefined) throw streamError;
}

/** Merges token deltas before they go out as events (§2.2). Every published
 *  event costs a seq, a stored row and a bus send; one per token is thousands
 *  a reply. Consecutive text (or reasoning) deltas are joined into one delta
 *  of the same shape, so a client reads them exactly as before. A held delta
 *  goes out after `windowMs`, or at once when any other kind of chunk comes,
 *  or on `flush`: the loop flushes before a step is committed, so its text is
 *  never published after the step that produced it. Sends go out in order. */
export class ChunkBatcher {
  private held: { type: string; textDelta: string } | null = null;
  private timer?: ReturnType<typeof setTimeout>;
  private chain: Promise<void> = Promise.resolve();

  constructor(
    private readonly send: (chunk: unknown) => Promise<void>,
    private readonly windowMs = 50,
  ) {}

  /** Take one chunk. A delta is held; anything else is sent, after what is
   *  held, and the promise resolves once it is out, so the step does not move
   *  on (a tool starting, say) before what it streamed is stored. */
  push(chunk: unknown): Promise<void> {
    const c = chunk as { type?: unknown; textDelta?: unknown } | null;
    const isDelta =
      (c?.type === 'text-delta' || c?.type === 'reasoning') && typeof c.textDelta === 'string';
    if (isDelta && this.held && this.held.type === c!.type) {
      this.held.textDelta += c!.textDelta as string;
      return Promise.resolve();
    }
    void this.flush();
    if (isDelta) {
      this.held = { ...(c as object), type: c!.type as string, textDelta: c!.textDelta as string };
      this.timer = setTimeout(() => void this.flush(), this.windowMs);
      (this.timer as unknown as { unref?: () => void }).unref?.();
      return Promise.resolve();
    }
    this.enqueue(chunk);
    return this.chain;
  }

  /** Send what is held, and resolve once everything pushed so far is out. */
  flush(): Promise<void> {
    if (this.timer) clearTimeout(this.timer);
    this.timer = undefined;
    const held = this.held;
    this.held = null;
    if (held) this.enqueue(held);
    return this.chain;
  }

  private enqueue(chunk: unknown) {
    this.chain = this.chain.then(() => this.send(chunk)).catch((err) => {
      // A lost live chunk is not a lost step: the step's messages are stored
      // at commit. It must not stop the chunks after it either.
      console.error('chunk not published', err);
    });
  }
}

/** A chunk as it is published: the SDK part, except that a tool result goes
 *  out in the shape the Go runtime sends, `{type, toolCallId, toolName,
 *  result}`, with `{error}` as the result of a call whose tool failed (the
 *  model got `error: <message>`, but a client marks the call failed from
 *  this), and a parked call's placeholder is not sent at all (its verdict
 *  arrives with the resume). Null means "do not publish". */
export function chunkPayload(chunk: unknown, toolErrors?: Map<string, string>): unknown | null {
  const c = chunk as { type?: string; toolCallId?: string; toolName?: string; result?: unknown } | null;
  if (c?.type !== 'tool-result') return chunk;
  const r = c.result as Record<string, unknown> | null;
  if (r && typeof r === 'object' && '__hitl_parked__' in r) return null;
  const failed = c.toolCallId ? toolErrors?.get(c.toolCallId) : undefined;
  if (failed !== undefined) toolErrors!.delete(c.toolCallId!);
  return {
    type: 'tool-result',
    toolCallId: c.toolCallId,
    toolName: c.toolName,
    result: failed !== undefined ? { error: failed } : c.result,
  };
}
